lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,298 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { BaseRequest, BaseResponse } from "../../framework";
+import OverrideableBuilder from "supertokens-js-override";
+import { GeneralErrorResponse, JSONObject } from "../../types";
+import { SessionContainer } from "../session";
+
+export type TypeInput = {
+    override?: {
+        functions?: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis?: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type TypeNormalisedInput = {
+    override: {
+        functions: (
+            originalImplementation: RecipeInterface,
+            builder?: OverrideableBuilder<RecipeInterface>
+        ) => RecipeInterface;
+        apis: (originalImplementation: APIInterface, builder?: OverrideableBuilder<APIInterface>) => APIInterface;
+    };
+};
+
+export type APIOptions = {
+    recipeImplementation: RecipeInterface;
+    config: TypeNormalisedInput;
+    recipeId: string;
+    isInServerlessEnv: boolean;
+    req: BaseRequest;
+    res: BaseResponse;
+};
+
+export type APIInterface = {
+    tokenPOST(input: {
+        clientId: string;
+        grantType: GrantType;
+        clientSecret?: string;
+        code?: string;
+        codeVerifier?: string;
+        refreshToken?: string;
+        redirectUri?: string;
+        scope: string[];
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<
+        | {
+              token_type: "Bearer";
+              access_token: string;
+              refresh_token?: string;
+              id_token?: string;
+              expires_in: number;
+          }
+        | GeneralErrorResponse
+    >;
+    // The OpenId spec states that we have to support both POST and GET
+    authorizeGET(input: {
+        clientId: string;
+        responseType: ResponseType[];
+        redirectUri: string;
+        scope: string[];
+        state?: string;
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: CodeChallengeMethod;
+
+        // OpenIdConnect
+        prompt?: OAuthPromptType;
+        nonce?: string;
+
+        session?: SessionContainer;
+
+        options: APIOptions;
+        userContext: any;
+    }): Promise<
+        | {
+              redirectUrl: string; // This will end up as a 302, encoding both error and success
+          }
+        | GeneralErrorResponse
+    >;
+    authorizePOST(input: {
+        clientId: string;
+        responseType: ResponseType[]; // add enum
+        redirectUri: string;
+        scope: string[];
+        state?: string;
```

Comment:
how is this input going to be used? The core does not have this as input / output anywhere

Reply:
It's just reflected redirection url. No need to store it.